lib/aws/request.js

```diff
@@ -0,0 +1,234 @@
+      }
+      return false;
+    };
+    // we translate params to an object for the service creation by selecting keys of interest
+    if (service.name === 'S3') {
+      // we need to take s3 acceleration options into account
+      const serviceParams = {
+        credentials: service.params.credentials,
+        useAccelerateEndpoint: shouldS3Accelerate(),
+        region: service.params.region ? service.params.region : 'us-east-1',
+      };
+      return new Service(serviceParams);
+    }
+    const serviceParams = {
+      credentials: service.params.credentials,
+      region: service.params.region ? service.params.region : 'us-east-1',
+    };
+    return new Service(serviceParams);
+  },
+  {
+    normalizer: ([service, method]) => {
+      return [deepSortObjectByKey(service), method].join('|');
+    },
+  }
+);
+
+/** Execute request to AWS service
+ * @param {Object} [service] - Description of the service to call
+ * @prop [service.name] - Name of the service to call, support subclasses
+ * @prop [service.params] - Parameters to apply when creating the service and doing the request
+ * @prop [service.params.credentials] - AWS Credentials to use
+ * @prop [service.params.useCache ] - Wether to reuse result of the same request cached locally
+ * @prop [service.params.region] - Region in which the call should be made (default to us-east-1)
+ * @prop [service.params.isS3TransferAccelerationEnabled] - Use s3 acceleration when available for the request
+ * @param {String} method - Method to call
+ * @param {Array} args - Argument for the method call
+ */
+function awsRequest(service, method, ...args) {
+  const BASE_BACKOFF = 5000;
+  const persistentRequest = (f) =>
+    new Promise((resolve, reject) => {
+      const doCall = (numTry) => {
+        f()
+          // We're resembling if/else logic, therefore single `then` instead of `then`/`catch` pair
+          .then(resolve, (e) => {
+            const { providerError } = e;
+            if (
+              numTry < MAX_RETRIES &&
+              providerError &&
+              ((providerError.retryable &&
+                providerError.statusCode !== 403 &&
+                providerError.code !== 'CredentialsError') ||
+                providerError.statusCode === 429)
+            ) {
+              const nextTryNum = numTry + 1;
+              const jitter = Math.random() * 3000 - 1000;
+              // backoff is between 4 and 7 seconds
+              const backOff = BASE_BACKOFF + jitter;
+
+              log(
+                [
+                  `Recoverable error occurred (${e.message}), sleeping for ~${Math.round(
+                    backOff / 1000
+                  )} seconds.`,
+                  `Try ${nextTryNum} of ${MAX_RETRIES}`,
+                ].join(' ')
+              );
+              setTimeout(doCall, backOff, nextTryNum);
+            } else {
+              reject(e);
+            }
+          });
+      };
+      return doCall(0);
+    });
```

Comment:
Let's construct this as:

```javascript
const persistentRequest = (f, numTry = 0) => f().catch(e => {
  ...
  return setTimeout(persistentRequest, backOff, f, nextNumTry);
});
```

Current promise construct is leaky and shares bad patterns

Reply:
Not sure about this one, I've tried an implementation, let me know if it matches what you had in mind :slightly_smiling_face: